src/meta/vab.c

```diff
@@ -0,0 +1,159 @@
+    wave_num = read_u16le(entry_off + 0x16, sf);
+
+    /* play default note */
+    pitch = SsPitchFromNote(60, 0, center, shift);
+
+    data_offset = is_vh ? 0x00 : (waves_off + 256 * 0x02);
+    for (j = 0; j < wave_num; j++) {
+        data_offset += read_u16le(waves_off + j * 0x02, sf) << 3;
+    }
+
+    data_size = read_u16le(waves_off + j * 0x02, sf) << 3;
+
+    /* build the VGMSTREAM */
+    vgmstream = allocate_vgmstream(1, 0);
+    if (!vgmstream) goto fail;
+
+    vgmstream->meta_type = meta_VAB;
+    vgmstream->coding_type = coding_PSX;
+    vgmstream->layout_type = layout_none;
+    vgmstream->sample_rate = (pitch * 44100) / 4096; // FIXME: Maybe use actual pitching if implemented.
+    vgmstream->num_samples = ps_bytes_to_samples(data_size, 1);
+    vgmstream->stream_size = data_size;
+    vgmstream->num_streams = tones;
+    snprintf(vgmstream->stream_name, STREAM_NAME_SIZE, "%02d/%02d", i, tone_num);
+
+    if (!vgmstream_open_stream(vgmstream, sf_data, data_offset))
+        goto fail;
+
+    if (is_vh) close_streamfile(sf_data);
+    return vgmstream;
+
+fail:
+    if (is_vh) close_streamfile(sf_data);
+    close_vgmstream(vgmstream);
+    return NULL;
+}
+
+/* Converts VAB note to PS1 pitch value (0-4096 where 4096 is 44100 Hz).
+ * Function reversed from PS1 SDK. */
+static uint16_t _svm_ptable[] =
+{
+    4096, 4110, 4125, 4140, 4155, 4170, 4185, 4200,
+    4216, 4231, 4246, 4261, 4277, 4292, 4308, 4323,
+    4339, 4355, 4371, 4386, 4402, 4418, 4434, 4450,
+    4466, 4482, 4499, 4515, 4531, 4548, 4564, 4581,
+    4597, 4614, 4630, 4647, 4664, 4681, 4698, 4715,
+    4732, 4749, 4766, 4783, 4801, 4818, 4835, 4853,
+    4870, 4888, 4906, 4924, 4941, 4959, 4977, 4995,
+    5013, 5031, 5050, 5068, 5086, 5105, 5123, 5142,
+    5160, 5179, 5198, 5216, 5235, 5254, 5273, 5292,
+    5311, 5331, 5350, 5369, 5389, 5408, 5428, 5447,
+    5467, 5487, 5507, 5527, 5547, 5567, 5587, 5607,
+    5627, 5648, 5668, 5688, 5709, 5730, 5750, 5771,
+    5792, 5813, 5834, 5855, 5876, 5898, 5919, 5940,
+    5962, 5983, 6005, 6027, 6049, 6070, 6092, 6114,
+    6137, 6159, 6181, 6203, 6226, 6248, 6271, 6294,
+    6316, 6339, 6362, 6385, 6408, 6431, 6455, 6478,
+    6501, 6525, 6549, 6572, 6596, 6620, 6644, 6668,
+    6692, 6716, 6741, 6765, 6789, 6814, 6839, 6863,
+    6888, 6913, 6938, 6963, 6988, 7014, 7039, 7064,
+    7090, 7116, 7141, 7167, 7193, 7219, 7245, 7271,
+    7298, 7324, 7351, 7377, 7404, 7431, 7458, 7485,
+    7512, 7539, 7566, 7593, 7621, 7648, 7676, 7704,
+    7732, 7760, 7788, 7816, 7844, 7873, 7901, 7930,
+    7958, 7987, 8016, 8045, 8074, 8103, 8133, 8162,
+    8192
+};
+
+static int16_t SsPitchFromNote(int16_t note, int16_t fine, uint8_t center, uint8_t shift) {
+
+    uint32_t pitch;
+    int16_t calc, type;
+    int32_t add, sfine;//, ret;
+
+    sfine = fine + shift;
+    if (sfine < 0) sfine += 7;
+    sfine >>= 3;
+
+    add = 0;
+    if (sfine > 15) {
+        add = 1;
+        sfine -= 16;
+    }
```

Comment:
The handing of sub-note fractions here doesn't look right.

First, note that `fine` and `shift` are unsigned range 0-127 (according to psx data format).

1. the interval between them is required to calculate output frequency scaling, so the expression should be `fine - shift` (requested fraction minus VAG center fraction), not `fine + shift`.  As coded, a VAG with non-zero center note shift will move the output frequency in the wrong direction (higher instead of lower).
2. the expression `if (sfine < 0) sfine += 7` is nonsensical because `fine + shift` will always be positive, and because the applied offset of 7 (or ~5.5 cents) has no musical meaning.

I think the correct handling would be something like:
```C
fine_interval = fine - shift;  // result range [-127..127]
add = 0;  // (better name might be note_offset)
if (fine_interval < 0) {
    fine_interval += 128;
    add = 1;
}
fine_interval >>= 3;  // quantize to 4 bits for table lookup
```

Reply:
I'm not sure about this, I'm looking at decompiled SsPitchFromNote function in Pac-Man World and it does do addition and not subtraction, then it takes the bottom 7 bits from result and adds 128 if bit 7 is set which cannot happen because it just chopped it off.

Reply:
If the PSX lib does do addition of the two 7 bit numbers (I can't think of why), certainly bit 7 can be set as a result.  In any case, the vgmstream code is not matching that because it assigns the addition to a signed int, then testing if the result is negative (which it will never be).  And nor is the vgmstream code adding 128, but 7 (which again corresponds to 5.5 cents of a music note, which doesn't make sense).

psx lib used lookup tables and all this silliness to avoid using float math.  vgmstream doesn't really need to do that, since the decoder is using floats anyway.

in a pure form, calculating the rate scale to play note X from a waveform corresponding to note Y:

```
scale = 2 ^ ((X - Y) / 12)
```
i.e. double the frequency for every octave X is higher than Y (or halve for every octave lover)

the cents or fine pitch is just the fractional part of X and Y.  For psx 7-bit form:
```
scale = 2 ^ (  (    (X + (float)X_fine/128) - (Y + (float)Y_fine/128)   ) / 12 )
```
note in the above that Y_fine (reference note) is subtracted from X_fine.

Reply:
Read my post again, it adds fine and shift and *then* takes the bottom 7 bits from the sum.

Reply:
> Read my post again, it adds fine and shift and _then_ takes the bottom 7 bits from the sum.

I understood it correctly.  That is effectively a modulo 128 after the addition.

But that is not related to `if (sfine < 0) sfine += 7`, which I've pointed out is nonsensical.  (Adding uchar's into an int will not be negative, and the magnitude 7/128 cents has no meaning in musical notation-- i.e. shifting the note by a fraction 0.055.)

Reply:
Like I said before, Sony code has the same issue, while it does 128 like you proposed, the value it checks cannot be negative at the point where the check is done so the check just always returns false, same thing as here.

Reply:
your original descirption:

> then it takes the bottom 7 bits from result and adds 128 if bit 7 is set which cannot happen because it just chopped that bit off.

I explained already, but this surmises wrong:  bit 7 (i.e. most significant bit, out of 0-7) can certainly become set after adding two 7-bit numbers.

What the psx lib is doing (add two 7-bit unsigned numbers into an 8-bit unsigned result and remove the MSB, which is a modulo of 128), is fundamentally different than what vgmstream is doing (add two 7-bit numbers into a 16-bit signed result, and check the sign i.e. bit 15).

Reply:
> then it takes the bottom 7 bits from result and adds 128 if bit 7 is set

Realized I may have misread this, sorry.  If you mean `b = a & 0x7F; if (b & 0x80) {...}` then that would be bizarre...

Reply:
I apologize, I misremembered slightly. What psx lib does is `c = a + b; c = c % 128; if (c & 0x8000) {...}`. c appears to be signed 16-bit integer and since integers are auto-promoted to 32-bit when arithmetic operation is performed, the result of that first operation cannot be negative. Then it does a modulo of 128 and checks bit 15 which can never be set unless fine value was negative.